### PR TITLE
add fake requests

### PR DESCRIPTION
**Pull request by ubamrein**
add random sleep
add tests

### CODE REVIEW

File: dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/controller/DPPPTController.java [hunk 1]
```diff
@@ -77,6 +79,7 @@ public DPPPTController(DPPPTDataService dataService, EtagGeneratorInterface etag
 		this.validateRequest = validateRequest;
 		this.batchLength = batchLength;
 		this.retentionDays = retentionDays;
+		this.random = new Random();
```
Comment (simonroesch) [inline dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/controller/DPPPTController.java:82]: use SecureRandom!

File: dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/security/JWTValidateRequest.java [hunk 2]
```diff
@@ -45,4 +45,17 @@ else if(request.getKeyDate() < OffsetDateTime.now().minusDays(21).toInstant().to
 		throw new IllegalArgumentException();
 	}
 
+	@Override
+	public boolean isFakeRequest(Object authObject, Object others) {
+		if (authObject instanceof Jwt) {
+			Jwt token = (Jwt) authObject;
+			boolean fake = false;
+			if (token.getClaimAsString("fake").equals("1")) {
```
Comment (simonroesch) [inline dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/security/JWTValidateRequest.java:53]: sollten wir hier nicht beide fake values anschauen? token UND json?

Comment (ubamrein) [reply dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/security/JWTValidateRequest.java:53]: ich weiss nicht... ich bin hier mal fuer nur auf token vertrauen, weil token signiert ist und sicher nicht gefälscht werden kann...

**Comment by jeffallen** [comment]
Where is this "sleep if fake" behaviour documented? What stack is it defending against?

**Comment by ubamrein** [comment]
We do sleep in any case and not just if it is a fake request. With that we try to prevent that  one can infer, if a request was fake or not just by timing the request.

File: dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/controller/DPPPTController.java [hunk 3]
```diff
@@ -108,8 +111,11 @@ public DPPPTController(DPPPTDataService dataService, EtagGeneratorInterface etag
 		if(!this.validateRequest.isFakeRequest(principal, exposeeRequest)) {
 			dataService.upsertExposee(exposee, appSource);
 		} 
+		
+		long after = System.currentTimeMillis();
+		long duration = after - now;
 		try{
-			Thread.sleep(random.nextInt(600));
+			Thread.sleep(this.requestTime - duration);
```
Comment (bachmann-ubique) [inline dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/controller/DPPPTController.java:115]: sollte evt noch Math.max(this.requestTime - duration, 0) sein? Siehe zb auch https://stackoverflow.com/questions/50269871/thread-sleep-1

Comment (simonroesch) [reply dpppt-backend-sdk/dpppt-backend-sdk-ws/src/main/java/org/dpppt/backend/sdk/ws/controller/DPPPTController.java:115]: Unbedingt, sonst gibt es eine Exception wenn der DB insert zu lange dauert!

**Comment by dcremonini** [comment]
Hi, what is the reason of sending fake requests?

**Comment by ubamrein** [comment]
The client sends fake requests to prevent any leakage of infection by analysing traffic.

Since the payload of the POST and GET requests differ substantially, an evil ISP or similar could then use analysis of the traffic, to see which IPs used the POST request and hence could find out who is infected.

To prevent timing attacks we further normalize the request time.

**Comment by dcremonini** [comment]
Very interesting. Thank you.

**Pull request merged**